### PR TITLE
Add thunk/adapter for FFI Upcall on Z

### PR DESCRIPTION
**Pull request by dchopra001**
Signed-off-by: Dhruv Chopra <Dhruv.C.Chopra@ibm.com>

### CODE REVIEW

**Comment by dchopra001** [comment]
Marked as a draft because I need to confirm that a failure in the builds is not related to my changes, and some formatting changes need to be made.

**Comment by ChengJin01** [comment]
FYI: @tajila, @pshipton, @DanHeidinga

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+				if (fprIndex <= 6) {
+					thunkMem += STDY(thunkMem, fprIndex, 0, C_SP_RegisterNum, offsetToParamArea + (stackSlotIndex * 8));
+				}
+				fprIndex += 2;
+				stackSlotIndex += 1;
+				break;
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_SP:
+				if (sigArray[i].sizeInByte == 4) {
+					if (fprIndex <= 6) {
+						thunkMem += STEY(thunkMem, fprIndex, 0, C_SP_RegisterNum, offsetToParamArea + (stackSlotIndex * 8));
```
Comment (zl-wang) [inline runtime/vm/xz64/UpcallThunkGen.cpp:437]: this looks wrong:  since single-precision FP is represented in the left part of the register, you cannot jump over structOffset to return the address later on.

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 2]
```diff
@@ -0,0 +1,767 @@
...
+			if (structSize != 1 && structSize != 2 && structSize != 4 && structSize != 8) {
+					isPointerToStruct = true;
+			}
+			else
+			{
+					// structs that are passed by value are expected to be stored as right aligned in the 8 byte
+					// stack slot. The return address to the struct must point to the start of the struct. So for
+					// structs smaller than 8 that are passed by value we must increment the offset by the required amount
+					// to adjust the final return address.
+					structOffset = 8 - structSize;
```
Comment (zl-wang) [inline runtime/vm/xz64/UpcallThunkGen.cpp:596]: mentioned above:  you cannot handle FPs similarly as to integer arguments.

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 1 above)
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp:437]: The structOffset will be zero when the struct size is 1, 2, 4 or 8 bytes.

On zLinux only structs of size 1, 2, 4, and 8 can be passed by value in a register. This is usually done in a GPR. 

The only scenario where this is done in an FPR is if we have a pure float/double struct with a single element (i.e. structs of size 4 and 8 bytes).

So I don't think we will end up in a problematic situation. Unless I'm missing sometime else here?

Comment (zl-wang) [reply runtime/vm/xz64/UpcallThunkGen.cpp:437]: No. the point is you cannot handle struct-of-1/2/4 bytes the same way with FPR and GPR arguments. the reason is: FPR is only populated in the left and GPR is only populated on the right. Imagine a single pure float struct to be passed ... it is passed by value in FPR with left half register populated ... you store it back with STEY at offset 0 in the stack slot ... later on, you have a structOffset value of 4 ... the return address actually points at the right half of the slot.

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 2 above)
Comment (zl-wang) [reply runtime/vm/xz64/UpcallThunkGen.cpp:596]: for ALL_SP with a single float, this calculation is obviously having structOffset set to 4.

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 2 above)
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp:596]: Ah I see what you mean. It does look like I missed that corner case. I'll rework this when I push new changes with the formatting fixed.

I'll also be adding to the upcall tests to capture all the zLinux corner cases (either in this PR or I'll pass them on to Cheng) so hopefully we won't have any more of these.

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 1 above)
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp:462]: This has been fixed now. `structOffset` will be zero for FPR argument types (including structs passed by value in FPR).

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 3]
```diff
@@ -0,0 +1,767 @@
...
+					I_32 structSizeInBytes = sigArray[i].sizeInByte;
+					if (structSizeInBytes == 1 || structSizeInBytes == 2 || structSizeInBytes == 4 || structSizeInBytes == 8)
+					{
+						if (gprIndex <= 6) {
+							thunkMem += STG(thunkMem, gprIndex, 0, C_SP_RegisterNum, offsetToParamArea + (stackSlotIndex * 8));
+						}
+						gprIndex += 1;
+                        stackSlotIndex += 1;
+					}
+					else
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: Both of these should be fixed now.

**Comment by dchopra001** [comment]
This commit should pass all up call tests found in @ChengJin01's branch: https://github.com/ChengJin01/openj9/find/jep389_419_ffi_upcall_v4_struct_v2_outofline_mh_v9_fin_v8
under the `test/functional/Java18andUp/src/org/openj9/test/jep419/upcall/` directory.

Note however that the Memory Padding in `UpcallMHWithStructTests.java` is not compatible on zLinux. Removing the padding on zLinux should resolve the problem. Just leaving this as a note here.

**Comment by dchopra001** [comment]
I also did my own unit testing as I was building the solution to make sure I'm covering all cases. On zLinux the parameter can be stored in one of two call frames depending on the situation. So there are some tests with lots of arguments to test these scenarios. I've put those tests [here](https://github.com/dchopra001/ffiUpCallTests). The [readme](https://github.com/dchopra001/ffiUpCallTests#test-details) describes the test cases. @ChengJin01 if you think any of these tests would be worth adding to your existing tests then let me know and I can provide you with the necessary files.

**Review by r30shah — CHANGES_REQUESTED** [review]
@dchopra001  Flushing out the first round of review. I have so far looked at the code of functions to generate the instructions in detail, most of the things looks OK, some things I observe while using `MVC` instructions that I think needs to be checked out.

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 4]
```diff
@@ -0,0 +1,762 @@
...
+	instructionPtr[4] = ((0x0F & B2) << 4) | ((D2 & 0xF00) >> 8);
+	instructionPtr[5] = D2 & 0xFF;
+	return instructionSize;
+}
+
+static I_32
+generateStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea)
+{
+	I_32 totalInstructionSize = 0;
+	I_32 maxBytes = 255;
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp:158]: So in revisiting this area I have a couple of observations:

- I don't think the issue mentioned above created any functional issues because the displacement integer calculation accounted for this discrepancy. This probably explains why I didn't see any failures in my test case because I do have edge cases for this scenario. Anyway, you're right that we should set maxBytes to 256. I'll have this fixed in the next commit.

- There is however a bug if we try to copy a structure larger than 4096 bytes. This is because the displacement integer in the instruction is only 12 bits in length. So in scenarios where we have to copy more than 4096 bytes of data, I propose the following solution: 
  - Copy the structure using consecutive MVC instruction for the first 4096 bytes
  - Copy the remaining bytes using an MVC instruction in a loop

The above solution is a small extension of the current method, and I imagine we won't have many practical scenarios where this could cause a performance concern. 

An alternative is to consider MVCL instructions but there are limitations there as well with respect to the displacement integer and for now I don't think the setup work required for that is worth it.

[the remark this answers is not in the document]
Comment (r30shah) [reply runtime/vm/xz64/UpcallThunkGen.cpp:158]: Good catch about the 4096 bytes, 
Your proposed solution means we will use 16 MVCs and for the remaining bytes use `MCVL/MVCLE` . Means we will use 100 bytes of instructions with that. Apart from that, you have to emit instruction to increase both source and destination by 4096 and also load the remaining bytes in the register for MVCLE.
Another caveat with using `MVCLE` looking at the PoPs, you would have to also insert branch instruction to account for interruption in copying from source to destination to account for operation completion because of CPU determined number of bytes have been moved without reaching the end of the first operand in which case condition code would be set. 
We should gauge if this is beneficial or simple loop of MVC (MVC, update src, update dst, check residue is less than 256 and branch back).

[the remark this answers is not in the document]
Comment (r30shah) [reply runtime/vm/xz64/UpcallThunkGen.cpp:158]: Also I need to confirm this from the instruction documents but the comment in [1] suggests a very high threshold should be used for deciding MVCL. 

[1]. https://github.com/eclipse/omr/blob/e26f652a53fef350a52dd999f2b1ce9a422ab2f0/compiler/z/codegen/OMRTreeEvaluator.cpp#L12813-L12818

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 5]
```diff
@@ -0,0 +1,762 @@
...
+	I_32 R1 = 1;
+	I_32 R2 = 2;
+	I_32 C_SP_R15 = 15;
+	// Load R1 with the hiddenParameter address
+	totalInstructionSize += LG(instructionPtr, R1, 0, C_SP_R15, offsetToParameterArea);
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: I've added these here: 39be597c2b57137681a8c4d40d31df5fc384b62e

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: This routine is now split into two. One copies small structs using upto 16 MVC instructions, and the other one copies larger structs in a loop with 16 MVC instructions at a time. I believe I'm missing one more corner case here. I'll update this comment once I've addressed that.

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 6]
```diff
@@ -0,0 +1,762 @@
...
+	instructionPtr += 6;
+
+	I_32 numFullSlots = structSize / 255;
+	I_32 residue = structSize % 255;
+
+	for(int i = 0; i < numFullSlots; i++) {
+		// MVC D1(L,B1),D2(B2)
+		// The 'L' field is 0-based
+		totalInstructionSize += MVC(instructionPtr, i*maxBytes/* D1 */, maxBytes - 1/* L */, R1/* B1 */, i*maxBytes/* D2 */, R2/* B2 */);
+		instructionPtr += 6;
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: The naming here should be improved now to make the function call more readable. I've also improved the variable names in the helper routines and updated how `instructionPtr` as per your suggestion above.

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 7]
```diff
@@ -0,0 +1,922 @@
...
+        totalInstructionSize += LG(instructionPtr, destinationRegister, 0, R15, offsetToParameterArea);
+	// Load loop counter into R3
+	totalInstructionSize += IILF(instructionPtr + totalInstructionSize, R3, numFullSlots);
+
+	for (int i = 0; i < maxLargeCopies; i++) {
+		// MVC D1(L,B1),D2(B2)
+		// The 'L' field is 0-based
+		destinationDisplacementInt = i*maxMVCCopyLength;
+		currMVCLength = maxMVCCopyLength - 1;
+		sourceDisplacementInt = i*maxMVCCopyLength;
```
Comment (joransiu) [inline runtime/vm/xz64/UpcallThunkGen.cpp:322]: Formatting Nit.  Spaces around `*`.

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 8]
```diff
@@ -0,0 +1,922 @@
...
+		totalInstructionSize += MVC(instructionPtr + totalInstructionSize, destinationDisplacementInt, currMVCLength, destinationRegister, sourceDisplacementInt, sourceRegister);
+	}
+	//printf("generateStructCopyInstructions_0: numLargeCopies: %d, residue: %d, structSize: %d, offsetParameter: %d\n", numLargeCopies, residue, structSize, offsetToParameterArea);
+	totalInstructionSize += AGFI(instructionPtr + totalInstructionSize, destinationRegister, maxMVCDisplacement);
+	totalInstructionSize += AGFI(instructionPtr + totalInstructionSize, sourceRegister, maxMVCDisplacement);
```
Comment (joransiu) [inline runtime/vm/xz64/UpcallThunkGen.cpp:327]: From an AGI pipeline bypass perspective, It's better to adjust the internal pointers with `LA/LAY Rdest, maxMVCDisplacement(Rdest)`.

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 9]
```diff
@@ -0,0 +1,922 @@
...
+		if (sigArray[lastSigIdx].sizeInByte <= maxMVCDisplacement)
+			thunkMem += generateSmallStructCopyInstructions(thunkMem, (I_64)sigArray[lastSigIdx].sizeInByte, offsetToParamArea);
+		else
+			thunkMem += generateLargeStructCopyInstructions(thunkMem, (I_64)sigArray[lastSigIdx].sizeInByte, offsetToParamArea);
+	}
+	thunkMem += LAY(thunkMem, R15, 0, R15, frameSize);
+	// Now load R14 with the return address.
+	thunkMem += LG(thunkMem, R14, 0, R15, offsetToR14Slot);
+	// Then branch back
+	thunkMem += BASR(thunkMem, R14, R14);
```
Comment (r30shah) [inline runtime/vm/xz64/UpcallThunkGen.cpp:701]: We do not return after this point right? In this case, we should use `BCR`.

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 10]
```diff
@@ -0,0 +1,922 @@
...
+	I_32 destinationDisplacementInt = -1;
+	I_32 sourceDisplacementInt = -1;
+	I_32 currMVCLength = -1;
+
+	// Load R4 with the hiddenParameter address
+        totalInstructionSize += LG(instructionPtr, destinationRegister, 0, R15, offsetToParameterArea);
+	// Load loop counter into R3
+	totalInstructionSize += IILF(instructionPtr + totalInstructionSize, R3, numFullSlots);
+
+	for (int i = 0; i < maxLargeCopies; i++) {
```
Comment (r30shah) [inline runtime/vm/xz64/UpcallThunkGen.cpp:317]: A comment or description on how the MVC loop works and residue is handled should be added here. Also is it possible to refactor the small struct copy and large struct copy and let single function handle both scenarios?

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 11]
```diff
@@ -0,0 +1,922 @@
...
+{
+	I_32 totalInstructionSize = 0;
+
+	const I_32 numFullSlots = structSize / maxMVCCopyLength;
+        const I_32 residue = structSize % maxMVCCopyLength;
+
+	const I_32 maxLargeCopies = 16;
+	const I_32 loopInstructionsLength = 114; // size in bytes
+	const I_32 sourceRegister = R2;
+	const I_32 destinationRegister = R4;
```
Comment (r30shah) [inline runtime/vm/xz64/UpcallThunkGen.cpp:307]: This means the function is fixed to use `R2` as source register and `R4` as destination register deeming it to be used in specific locations only. Should this be parameterized to allow it to be used if we ever have to copy with different source and destination register?

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 12]
```diff
@@ -0,0 +1,922 @@
...
+	Assert_VM_true((argIdx >= 0) && (argIdx < lastSigIdx));
+
+	// Test the return type
+	I_32 argSize = sigArray[argIdx].sizeInByte;
+
+	I_32 gprIndex = 1;
+	I_32 fprIndex = -2;
+	I_32 callerFrameIndex = -1; // values that weren't passed by register
+	I_32 calleeFrameIndex = -1; // values that were passed through register
+	bool argumentInCallerFrame = false; // At first we assume the argument hasn't overflown and is in the new frame allocated by the callee
```
Comment (r30shah) [inline runtime/vm/xz64/UpcallThunkGen.cpp:736]: Should be `overflowed`

**Review by r30shah — CHANGES_REQUESTED** [review]

**Comment by joransiu** [comment]
jenkins test sanity zlinux jdk8,jdk19

**Comment by r30shah** [comment]
@joransiu any concerns/review with this change? Since @dchopra001  have tested the changes with different test case, and the sanity test passes, should we attempt merging this one?

**Comment by joransiu** [comment]
> @joransiu any concerns/review with this change? Since @dchopra001 have tested the changes with different test case, and the sanity test passes, should we attempt merging this one?

I would like to see if @dchopra001 will get a chance to address the few minor outstanding comments, and squash the commits.   

@zl-wang / @ChengJin01 Do we have any specific tests we should trigger to exercises these paths?

**Comment by ChengJin01** [comment]
@joransiu,  @dchopra001 should be able to verify this PR with his own tests specific to Z at https://github.com/dchopra001/ffiUpCallTests.  In terms of generic tests suites for upcall, our test cases at https://github.com/eclipse-openj9/openj9/pull/15310 have not yet merged to the repo as we have not yet finished the code review for all PRs listed at https://github.com/eclipse-openj9/openj9/issues/15068. So there is no way to directly verify with these test suites for the moment due to the incomplete code.

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 12 above)
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp:730]: This should be fixed now.

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 11 above)
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: Now the source and destination register operands are parametrized for the struct copy routines: 109cbe6452e154afe9686653d3693605391b8423

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 9 above)
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: This should be fixed in: c54d5bd7f1575e9af8aa7f5c2fd37c396ce8be05

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 13]
```diff
@@ -0,0 +1,762 @@
...
+generateStructCopyInstructions(I_8 *instructionPtr, I_32 structSize, I_32 offsetToParameterArea)
+{
+	I_32 totalInstructionSize = 0;
+	I_32 maxBytes = 255;
+	I_32 R1 = 1;
+	I_32 R2 = 2;
+	I_32 C_SP_R15 = 15;
+	// Load R1 with the hiddenParameter address
+	totalInstructionSize += LG(instructionPtr, R1, 0, C_SP_R15, offsetToParameterArea);
+	instructionPtr += 6;
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: Should be fixed now as I mentioned on the comment on line 172

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 14]
```diff
@@ -0,0 +1,922 @@
...
+	const I_32 offsetToR14Slot = 112;
+
+	// First extend the stack frame
+	// Save the original return address into memory
+	thunkMem += STG(thunkMem, R14, 0, R15, offsetToR14Slot);
+	// Store the current C stack pointer at the first byte of the new frame.
+	thunkMem += STG(thunkMem, R15, 0, R15, -frameSize);
+	// Then update R15 with the new C stack pointer.
+	// Now do an add to update the stack pointer in R15.
+	thunkMem += AGFI(thunkMem, R15, -frameSize);
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: LAY instructions being used here: 68ef0623c789bfab3d1dcff2dc46fee881686cbf

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 15]
```diff
@@ -0,0 +1,922 @@
...
+			default:
+				//printf("createUpCallThunk_2 Can't find argument type: %d\n", sigArray[i].type);
+				Assert_VM_unreachable();
+		}
+	}
+
+	//printf("createUpCallThunk_3 - gprArgCount: %d, fprArgCount: %d, numStackSlots: %d, numInstructions: %d lastSigIndex: %d\n", gprArgCount, fprArgCount, numStackSlots, numInstructions, lastSigIdx);
+
+	I_32 frameSize = 0;
+	const I_32 offsetToParamArea = 160;
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp:544]: I added some comments here: 6e50a19f91bf9b540b7fd894d891a7b8a9f08aa3

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 16]
```diff
@@ -0,0 +1,922 @@
...
+	I_32 roundedCodeSize = 0;
+	I_8 *thunkMem = NULL;
+
+	// Stack space to store arguments or return values
+	frameSize = offsetToParamArea + numStackSlots * 8;
+
+	// 8 byte align the thunk memory.
+	// Fixed amount of instructions needed for backing up and restoring data before/after we invoke the common dispatcher
+	const I_32 extraInstructionsSize = 64;
+	roundedCodeSize = ((( numInstructions * 6 + extraInstructionsSize) / 8) + 1) * 8;
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: This should be fixed now.

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 17]
```diff
@@ -0,0 +1,922 @@
...
+			// be found in the caller's frame
+			hiddenParameter = true;
+			// extra instruction used to store hidden parameter in callee frame
+			numInstructions++;
+			// add an extra slot to callee's frame to store the hidden parameter
+			numStackSlots += 1;
+			// We will use MVC to copy over the struct bytes. The structSize / 256 tells us how many MVC instructions we need
+			// in a loop. Then we add 1 to account for any residual bytes.
+			if (sigArray[lastSigIdx].sizeInByte <= maxMVCDisplacement)
+				numInstructions += 2 + sigArray[lastSigIdx].sizeInByte / maxMVCCopyLength;
```
[the remark this answers is not in the document]
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: Added comments around how this is calculated and updated this area as well: 193521ffdae93a7ae820818ab09473e47b0b2a9b

**Comment by ChengJin01** [comment]
@dchopra001 & @joransiu, could this PR be merged if there is no more concern to be addressed?

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 18]
```diff
@@ -0,0 +1,930 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2022 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available undeMAX_MVC_COPY_LENGT * Secondary Licenses when the conditions for such availability set
```
Comment (r30shah) [inline runtime/vm/xz64/UpcallThunkGen.cpp:10]: Seems like some kind of mixed up here.

**Review by r30shah — CHANGES_REQUESTED** [review]
Minor mixed up in the license header text, apart from that, I am OK with most of the changes made by @dchopra001. @ChengJin01  What is the timeline within which we need to get this change in?

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 18 above)
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: Should be fixed here: 565c53270b39f2e5b6923add77381ea9975c9336

**Comment by ChengJin01** [comment]
> @ChengJin01 What is the timeline within which we need to get this change in?

@r30shah,  we should have sufficient time for this PR but I'd like to have it merged as soon as possible if we could. So I will need to double-check with jtreg tests to see whether there is any platform specific issue to be addressed on zLinux.

**Comment by dchopra001** [comment]
I did test the last time I pushed code changes to this branch a week or two ago. Just to be sure I'm doing a clean build and will test again. I'll update here once the build is done and the tests are clear.

File: runtime/vm/xz64/UpcallThunkGen.cpp [hunk 19]
```diff
@@ -0,0 +1,931 @@
...
+	// loopInstructionsLength = 16x MVC(6) + 2x LAY (6) + 1x AFI (6)
+	const I_32 loopInstructionsLength = 114; // size in bytes
+	I_32 destinationDisplacementInt = -1;
+	I_32 sourceDisplacementInt = -1;
+	I_32 currMVCLength = -1;
+
+	// Load loop counter into R3
+	totalInstructionSize += IILF(instructionPtr + totalInstructionSize, R3, numFullSlots - loopResidue);
+
+	// Generate a loop that will copy 4096 bytes at a time (i.e. 16 MVC instructions). The loop concludes where there
```
Comment (joransiu) [inline runtime/vm/xz64/UpcallThunkGen.cpp:311]: `The loop concludes where...` may be better worded as `The loop concludes when...`

**Review by joransiu — APPROVED** [review]
LGTM.  One minor comment nit.

**Comment by joransiu** [comment]
Not sure why the Copyright check and Line Ending checks suddenly flagged issues in other files, and shows conflicts on merge.   @dchopra001 , can you squash and rebase?

@zl-wang / @r30shah : Any other comments/feedback?

**Comment by ChengJin01** [comment]
@AdamBrousseau, is there any update in terms of the Copyright & Line Endings Check ? It seems the failing lines listed in there totally have nothing to do with this PR.

**Comment by AdamBrousseau** [comment]
@ChengJin01 the checks don't work properly when there are merge conflicts in the PR. Once they are resolved, the check should return proper results.

**Comment by ChengJin01** [comment]
@dchopra001, please double-check whether there is any conflict to be resolved in this PR.

**Comment by r30shah** [comment]
> @zl-wang / @r30shah : Any other comments/feedback?

Waiting for @dchopra001 's comment for verifying testing. Overall, changes looks good to me.

File: runtime/vm/xz64/UpcallThunkGen.cpp (hunk 19 above)
Comment (dchopra001) [reply runtime/vm/xz64/UpcallThunkGen.cpp]: This should now be fixed in the final squashed commit.

**Comment by dchopra001** [comment]
Merge conflicts resolved and all my local tests have passed as well.

**Review by r30shah — APPROVED** [review]

**Comment by joransiu** [comment]
jenkins test sanity zlinux jdk8,jdk19

**Review by zl-wang — APPROVED** [review]
Looks good to me
